### PR TITLE
JNISystemDependency fixes

### PR DESCRIPTION
**Pull request by tristan957**
First commit is a cleanup commit.

Second commit will fix https://github.com/mesonbuild/meson/issues/11173.

### CODE REVIEW

**Comment by tristan957** [comment]
cc @eli-schwartz

We will have to see if the macOS tests break.

**Comment by bgilbert** [comment]
Note that Java integration tests are currently disabled on macOS:

https://github.com/mesonbuild/meson/blob/dd25b88f009ee47579b33dfbd167a9434cc8f2be/run_project_tests.py#L1094

**Comment by tristan957** [comment]
I guess if you say that it works, then we're good to go.

**Comment by bgilbert** [comment]
On my system this works as designed:

```console
$ meson setup build
The Meson build system
Version: 1.0.0.rc1
Source dir: /Users/bgilbert/report
Build dir: /Users/bgilbert/report/build
Build type: native build
Project name: Test
Project version: undefined
C compiler for the host machine: cc (clang 12.0.0 "Apple clang version 12.0.0 (clang-1200.0.32.29)")
C linker for the host machine: cc ld64 609.8
Java compiler for the host machine: javac (unknown 19.0.1)
Host machine cpu family: x86_64
Host machine cpu: x86_64
ERROR: Darwin-based systems, like macOS, provide stubs for various Java binaries. These
binaries are provided in directories that don't match the heuristics of a
typical JDK installation. The JNI system dependency relies on these heuristics
to determine the location of the JNI headers.

To work around this issue, set the JAVA_HOME environment variable to the output
of /usr/libexec/java_home.

See https://stackoverflow.com/a/15133344/7572728 for more details.

Run-time dependency jni found: NO (tried system)

meson.build:7:0: ERROR: Dependency "jni" not found, tried system

A full log can be found at /Users/bgilbert/report/build/meson-logs/meson-log.txt
$ JAVA_HOME=$(/usr/libexec/java_home) meson setup build
The Meson build system
Version: 1.0.0.rc1
Source dir: /Users/bgilbert/report
Build dir: /Users/bgilbert/report/build
Build type: native build
Project name: Test
Project version: undefined
C compiler for the host machine: cc (clang 12.0.0 "Apple clang version 12.0.0 (clang-1200.0.32.29)")
C linker for the host machine: cc ld64 609.8
Java compiler for the host machine: javac (unknown 19.0.1)
Host machine cpu family: x86_64
Host machine cpu: x86_64
Run-time dependency jni found: YES 19.0.1
Has header "jni.h" with dependency jni: YES 
Build targets in project: 0

Found ninja-1.11.1 at /opt/local/bin/ninja
```

Though note that the first line of the error extends past 80 characters, since it doesn't account for the width of `ERROR: `.

However, this design seems surprising to me: the code knows how to fix the issue, but tells the user to do it, rather than doing so directly.  Wouldn't it be better to just run `/usr/libexec/java_home` and incorporate the result?

**Comment by tristan957** [comment]
Not sure what Meson usually does for message widths. Maybe I should be removing newlines.

What if the javac used by java_home isn't the same as the stub javac? Is it possible for them to match? I don't know. Maybe Meson should just reject this dumb javac in the project() call.

**Comment by tristan957** [comment]
Rereading the stack overflow answer, the stub seems to call out to whatever the default jdk is, so I assume that `java_home(1)` can't be out of sync with the default jdk. I can probably just reach out to the `/usr/libexec/java_home` executable and make it work magically on Darwin systems. I will take another crack tonight.

**Comment by tristan957** [comment]
@bgilbert please `pull --rebase` my branch and try again.

File: mesonbuild/dependencies/dev.py [hunk 1]
```diff
@@ -546,6 +549,15 @@ def __init__(self, environment: 'Environment', kwargs: JNISystemDependencyKW):
         self.java_home = environment.properties[self.for_machine].get_java_home()
         if not self.java_home:
             self.java_home = pathlib.Path(shutil.which(self.javac.exelist[0])).resolve().parents[1]
+            if m.is_darwin():
+                problem_java_prefix = pathlib.Path('/System/Library/Frameworks/JavaVM.framework/Versions')
+                if problem_java_prefix in self.java_home.parents:
+                    res = subprocess.run(['/usr/libexec/java_home', '--failfast', '--arch', m_cpu_family],
```
Comment (tristan957) [inline mesonbuild/dependencies/dev.py:555]: @eli-schwartz any thoughts on what if anything should be logged if this command fails?

Comment (eli-schwartz) [reply mesonbuild/dependencies/dev.py:555]: "Your binary is not an allowed client of /usr/libexec/java_home", and leave it to the user to guess whether the binary is javac or Meson.

We want macOS users to feel at home, after all. Plus it's good to follow Apple's UX guidelines.

**Comment by bgilbert** [comment]
It fails on my system:

```console
$ ~/sw/meson/env/bin/meson setup build
The Meson build system
Version: 1.0.0.rc1
Source dir: /Users/bgilbert/report
Build dir: /Users/bgilbert/report/build
Build type: native build
Project name: Test
Project version: undefined
C compiler for the host machine: cc (clang 12.0.0 "Apple clang version 12.0.0 (clang-1200.0.32.29)")
C linker for the host machine: cc ld64 609.8
Java compiler for the host machine: javac (unknown 19.0.1)
Host machine cpu family: x86_64
Host machine cpu: x86_64
Unable to find any JVMs matching architecture "amd64".
ERROR: JAVA_HOME could not be discovered on the system. Please set it explicitly.
Run-time dependency jni found: NO (tried system)

meson.build:7:0: ERROR: Dependency "jni" not found, tried system

A full log can be found at /Users/bgilbert/report/build/meson-logs/meson-log.txt
```

It's passing `--arch amd64` to `/usr/libexec/java_home`, which is invalid.  I've verified that `m_cpu_family` is set to `amd64`.  When I hardcode the `--arch` argument to `x86_64`, it works correctly.  I'm using this `meson.build`:

```
project(
  'Test',
  'c', 'java',
  meson_version : '>=0.62',
)

jni = dependency('jni')
meson.get_compiler('c').has_header(
  'jni.h',
  dependencies : [jni],
  required : true,
)
```

With the `x86_64` hardcoding in place, the Java integration tests are also passing on my machine, so you may wish to drop the macOS Java exclusion from `run_project_tests.py`.

**Comment by tristan957** [comment]
Ok. Thanks for the feedback. I'll remove the test restrictions. This x86_64 vs amd64 is very annoying. Easy fix though.

**Comment by bgilbert** [comment]
Thanks for the fix!

**Pull request merged**